g2p_programs/models/managers/deduplication_manager.py

```diff
@@ -174,21 +171,86 @@ class IDDocumentDeduplication(models.Model):
-            # duplicates
-            _logger.info("Deduplicate beneficiaries: %s", beneficiaries)
+        self.ensure_one()
+        program = self.program_id
+        beneficiaries = program.get_beneficiaries(states)
+        sql = """
+            SELECT * FROM
+            (SELECT g2p_reg_id.id as identity_id,
+                id_type,
+                value,
+                g2p_reg_id.partner_id as identity_partner_id,
+                g2p_group_membership.group as identity_group_id,
+                count(*)
+            OVER
+                (PARTITION BY
+                    id_type,
+                    value
+                ) AS count
+            FROM g2p_reg_id
+            -- Join to the group it belong to through the individual
+            LEFT join g2p_group_membership ON g2p_group_membership.individual = g2p_reg_id.partner_id
+            -- Join the group to the program
+            LEFT join g2p_program_membership as group_program_membership
+                ON group_program_membership.partner_id = g2p_group_membership.group
+            -- Join the registrant to the program
+            LEFT join g2p_program_membership as program_membership
+                ON program_membership.partner_id = g2p_reg_id.partner_id
+            where
+            (
+                group_program_membership.program_id = %s OR
+                program_membership.program_id = %s
+            ) AND (
+                group_program_membership.state IN %s OR
+                program_membership.state IN %s
+            ) AND (
+                g2p_reg_id.expiry_date IS NULL OR
+                g2p_reg_id.expiry_date > CURRENT_DATE
+            )
+            ) tableWithCount
+            WHERE tableWithCount.count > 1;"""
+
+        states = (*states,)
+        self.env.cr.execute(sql, (program.id, program.id, states, states))
+        duplicates = self.env.cr.dictfetchall()
+        duplicate_partner_ids = []
+        for rec in duplicates:
+            _logger.info("DEBUG: duplicate: %s", rec)
             if program.target_type == "group":
-                duplicate_beneficiaries = (
-                    self._check_duplicate_by_group_with_individual(beneficiaries)
-                )
+                if rec["identity_group_id"]:
+                    duplicate_partner_ids.append(rec["identity_group_id"])
+                else:
+                    duplicate_partner_ids.append(rec["identity_partner_id"])
             else:
-                duplicate_beneficiaries = self._check_duplicate_by_individual(
-                    beneficiaries
-                )
-            return len(duplicate_beneficiaries)
+                duplicate_partner_ids.append(rec["identity_partner_id"])
+
+        _logger.info("DEBUG: duplicate partner ids: %s", duplicate_partner_ids)
+
+        duplicate_beneficiaries = beneficiaries.filtered(
+            lambda rec: rec.partner_id.id in duplicate_partner_ids
+        )
```

Comment:
We are losing all the benefits of the optimization we did with the SQL query earlier. 
There could be millions of records in a program we cannot be filtering in odoo, it needs to be done in the DB

Reply:
I have made some fixes based on your comments. Please check again.